### PR TITLE
feat(feature dev): Filter out non-code files at preparation state

### PR DESCRIPTION
**Pull request by willyyhuang**
Feature dev compiles a compressed file with relevant local files to prepare response. This process would allow some irrelevant files to be added into the zip without many filters.

The current filtering pattern: ignore the gitignores.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
Implement an allowlist behaviour checking file extensions when preparing repository for Q feature dev.

Filter out any files that aren't in the allowlist

Tested by building plugin locally, verified the s3 artifact size, also added unit test to reflect this behaviour.

Waiting for this PR first: https://github.com/aws/aws-toolkit-jetbrains/pull/4685

Before
image

After
image

Example telemetry event to be emitted:
```
2024-07-15 15:15:24,382 [  51550]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: null, count: 1
2024-07-15 15:15:24,388 [  51556]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: xml, count: 6
2024-07-15 15:15:24,389 [  51557]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: gitignore, count: 5
2024-07-15 15:15:24,389 [  51557]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: iml, count: 1
2024-07-15 15:15:24,389 [  51557]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: yaml, count: 4
2024-07-15 15:15:24,389 [  51557]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: graphql, count: 13
2024-07-15 15:15:24,389 [  51557]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: nvmrc, count: 2
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: info, count: 1
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: png, count: 2
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: editorconfig, count: 1
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: eslintignore, count: 2
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: yml, count: 1
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: prettierignore, count: 2
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: ownership-file, count: 1
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: zip, count: 3
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: out, count: 11
2024-07-15 15:15:24,390 [  51558]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: dot, count: 3
2024-07-15 15:15:24,391 [  51559]   INFO - software.aws.toolkits.jetbrains.services.amazonq.FeatureDevSessionContext - ignored filename: map, count: 32
```

## Checklist
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [x] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (tverney) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:69]: nit - can we potentially load this from a json file so we don't have a long kt code?

File: .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json
Comment (tverney) [inline .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: by relevant code you mean code extensions/languages?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (tverney) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:277]: just curious about the TODO: there is any blocker for we do right now?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (tverney) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:303]: do we need return this value?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:69]: hm, potentially, however I do not see that as a pattern in this package

File: .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json
Comment (willyyhuang) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: These messages are supposed to be non-technical, therefore it is phrased that way

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:277]: I believe telemetry definition has to be defined first then can be emitted, the definition is not yet defined

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:303]: Yes, otherwise the filtering logic won't work

File: .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json
Comment (tverney) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: I see. "relevant code" sounds a bit ambiguous to me. Not sure if "code extensions" is technical in this context.

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (tverney) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:69]: yes, but I would argue that I don't see another large string like this one set in the package that requires extract to a file. Maybe would be good do for this one?

File: .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json
Comment (willyyhuang) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: since the VSCode equivalent has the change log, I don't think it's worth changing as they reflect the same change: packages/amazonq/.changes/next-release/Feature-dec3e22a-a3f1-4bbc-ae69-a564474560e1.json

Comment (tverney) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: I see. 

So from what you've said here we're not changing something because the same thing its already in a commit in VS Code?

Well, that's fine. I get your efficiency point.

My idea here is try to look in user's perspective and understand if the textual makes sense. I would say that its fair enough that users would like to know that we're evaluating their code extensions and this ideally should be self-explanatory in the changelog. 

In the other hand, I would have an immediate question in my mind to understand what's a "relevant code" in our perspective, which might sounds imperious - are we defining what is relevant or not (?) and leaving the user out of the discussion?

Comment (willyyhuang) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: > So from what you've said here we're not changing something because the same thing its already in a commit in VS Code?

> Well, that's fine. I get your efficiency point.

No, I am open to change the change log texts as well, it's not about efficiency, please give me any other suggested change log texts.

However, here is a guidance for change log: https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#changelog

Based on this, I would say that the current description satisfies all of the points listed, if the user is interested in technical details, then checking out a commit is not a problem as this is an open source project.

Comment (tverney) [reply .changes/next-release/feature-a56afd04-cf25-4417-97ab-298555dce65a.json:3]: Well, I don't see as intuitive an user clicking in a commit when its seeing the changelog in the IDE installing the latest version of Amazon Q in VS Code.

Yes, my suggestion:

"Q feature /dev: Use common code extensions to filter related relevant files"

**Review by tverney — APPROVED** [review]

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt
Comment (chengoramazon) [inline plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt:219]: Curious: why did this get changed?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (chengoramazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:70]: +1, json or a constant file would make the actual code much readable.

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt
Comment (willyyhuang) [reply plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt:219]: this is because the CodeTransform team changed the event name in the newer version of `aws-toolkit-common`, but it is not updated in the IDE toolkit code yet, without changing this, the build would fail

VSCode equivalent of this PR also has the same change: https://github.com/aws/aws-toolkit-vscode/pull/5296/files

**Review by chengoramazon — APPROVED** [review]

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt
Comment (tincheng) [reply plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformTelemetryManager.kt:219]: The change to the metric was not by code transform team. We are adding back the metric that was removed. Please do not change the usage of this metric. https://github.com/aws/aws-toolkit-common/pull/788

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:12]: need ack from /transform and chat that this will not affect their feature

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:113]: if `VirtualFile#getExtension` exists, some of this seems unnecessary

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:130]: isnt extension always a string?

Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:130]:
```kt
val ignoredExtensionMap = mutableMapOf<String, Int>().withDefault { 0 }
...
ignoredExtensionMap[file.extension] += 1
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:84]: `VirtualFile#isDirectory` exists

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:139]: aren't these binary files?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:198]: do we want to reference the source? this list seems very arbitrary and covers many obscure languages

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:84]: Let me refactor this, there is no need to cast to `File` type

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:198]: I will reference the list in the next update, thanks for pointing it out

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:130]: Since `file.extension` is nullable
```
public @Nullable @NlsSafe String getExtension()
```

This will cause unresolved reference

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:113]: agreed, will refactor

**Comment by willyyhuang** [comment]
FYI: there is a defect in `PluginUpdateManager.kt` post telemetry definition generation, due to a change in schema, waiting for code owner to apply an update.

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:35]: use the kotlin extensions
```suggestion
        whenever(txtFile.extension).thenReturn("txt")
```

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:33]:
```suggestion
        txtFile = mock<VirtualFile>()
```

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:26]: these are all used in a single test so they dont really need to be initialized in setUp()

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:105]: just short circuit

```suggestion
        val extension = file.extension ?: return false
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt]: deleted in new rev

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:13]: there is no overlap between this change and gumby's file zipping logic

Comment (tincheng) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:13]: @rli can you share why you think this will affect /transform and chat?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:88]: these are also binary

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:219]: `.yml` is also valid for yaml

**Review by rli — APPROVED** [review]

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (tverney) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:219]: +1

**Review by tverney — APPROVED** [review]

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:13]: in `CodeScanSessionConfig.kt`, it is re-using `ignoreFile` func rather than `zipFiles`, so I don't think this change will affect Gumby's functionality

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt
Comment (willyyhuang) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QConstants.kt:219]: as addressed previously, I will create a follow up PR to sync and re-evaluate the file extension constants

**Pull request merged**
